RavenDB-13041 Elasticsearch ETL

arekpalinski opened the pull request:
https://issues.hibernatingrhinos.com/issue/RavenDB-13041

djordjedjukic committed:
RavenDB-13041 Elasticsearch ETL

RavenDB-13041 fixing compilation

rename Elastic to Elasticsearch


implement CreateScope method


modify RemoveConnectionString method


remove unused code


add missing part in license status


PR comments resolved

djordjedjukic committed:
RavenDB-13041 Elasticsearch ETL

RavenDB-13041 fixing compilation

rename Elastic to Elasticsearch


implement CreateScope method


modify RemoveConnectionString method


remove unused code


add missing part in license status


PR comments resolved

arekpalinski commented:
Originally reviewed in the following PR: https://github.com/arekpalinski/ravendb/pull/4

ayende commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,62 @@
+{
+    public class ElasticsearchConnectionString : ConnectionString
+    {
+        public string[] Nodes;
```
How does authentication is expected to work here?

djordjedjukic replied:
I need to think about this. I will back to you with the answer.

djordjedjukic replied:
There are few options for authentication.

- ApiKeyAuthentication
- BasicAuthentication
- ClientCertificate
- ClientCertificates

https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/configuration-options.html

Should we support all of them or just some of them? @ayende

arekpalinski replied:
I'm for ApiKey and Basic auth support

ayende replied:
I would rather support them all and let the user decide.

Need to ensure that for secured database, we raise a warning if they are sending to an insecured Elastic isntance.

arekpalinski replied:
1. I wanted to avoid keeping the certificate but OK - we can keep is as base64. 
2. Currently we rise this warning when Elasticsearch instance doesn't have "https" (just "http") in URL - so same logic as with RavenDB. But indeed, we should also warn if there is no authentication required.

arekpalinski replied:
@djordjedjukic 

Regarding point 2 - I think the best place to check it and warn if needed would be here:

https://github.com/ravendb/ravendb/blob/ab4da6d7416e0174d483f8064d6506a5852ce58c/src/Raven.Server/Documents/ETL/EtlLoader.cs#L333-L334

then you could do

```
if ((_databaseRecord.Encrypted && config is ElasticSearchConfig esConfig && esConfig.Connection.Auth == null)
{
   LogConfigurationWarning()
}

```

arekpalinski replied:
Note that we don't want to return `false` - the config is still valid but just want to warn a user

djordjedjukic replied:
done

ayende commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchEtlConfiguration.cs:
```diff
@@ -0,0 +1,62 @@
+        public string IndexName { get; set; }
+        public string IndexIdProperty { get; set; }
+
+        protected bool Equals(ElasticsearchIndex other)
```
Missing `GetHashCode` ? Better to just use a `record`, no?

djordjedjukic replied:
I agree. Working on that.

djordjedjukic replied:
The equals method is removed because it was never used. Also, records aren't used anywhere in the project so I wasn't sure if that's going to break something.

ayende commented on src/Raven.Client/Documents/Smuggler/SmugglerResult.cs:
```diff
@@ -407,6 +417,12 @@ public override string ToString()
+                    sb.AppendLine("- ELASTICSEARCH Connection Strings");
+
+                if (ElasticsearchEtlsUpdated)
+                    sb.AppendLine("- ELASTICSEARCH ETLs");
```
Why the caps?

djordjedjukic replied:
done

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1060,6 +1065,12 @@ private void ThrowIfCannotActivateLicense(LicenseStatus newLicenseStatus)
+            
+            if (elasticsearchEtlCount > 0 && newLicenseStatus.HasElasticsearchEtl == false)
+            {
+                var message = GenerateDetails(elasticsearchEtlCount, "ELASTICSEARCH ETL");
```
Why the caps?

ayende replied:
Not that `SQL ETL` or `OLAP ETL` - those are acronyms.
But `Raven ETL` is not in caps

djordjedjukic replied:
I agree. Working on that.

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Commercial/LicenseManager.cs:
```diff
@@ -1375,6 +1386,18 @@ public void AssertCanAddSqlEtl()
+            if (LicenseStatus.HasElasticsearchEtl != false)
+                return;
+
+            const string message = "Your current license doesn't include the Elasticsearch ETL feature";
```
```suggestion
            const string message = "Your current license doesn't include the ElasticSearch ETL feature";
```

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -229,6 +240,15 @@ private void HandleChangesSubscriptions()
                         else
                             connectionStringNotFound = true;
                         break;
+                    case EtlType.Elasticsearch:
```
And in general, let's use `ElasticSearch` and not `Elasticsearch`

djordjedjukic replied:
The first version was just `Elastic`, then we switched to `Elasticsearch` because they are using that phrase on their website.
But I will change that.

Danielle9897 replied:
Using `ElasticSearch` was my initial reaction as well...
But since the formal name they use all over is `Elasticsearch` then we probably better follow that...

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchDocumentTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+
+        protected override void AddLoadedAttachment(JsValue reference, string name, Attachment attachment)
+        {
+            throw new NotSupportedException("Attachments aren't supported by ELASTICSEARCH ETL");
```
caps (and elsewhere in the file)

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+        {
+        }
+
+        public const string ElasticsearchEtlTag = "ELASTICSEARCH ETL";
```
```suggestion
        public const string ElasticSearchEtlTag = "ElasticSearch ETL";
```

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+        protected override IEnumerator<ElasticsearchItem> ConvertAttachmentTombstonesEnumerator(DocumentsOperationContext context, IEnumerator<Tombstone> tombstones,
+            List<string> collections)
+        {
+            throw new NotSupportedException("Attachment tombstones aren't supported by ELASTICSEARCH ETL");
```
Caps (and elsewhere in the file)

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+using System.Linq;
+using System.Text;
+using Elasticsearch.Net;
+using Nest;
```
What is this?

djordjedjukic replied:
`NEST` is the official high-level .NET client of Elasticsearch.
`Elasticsearch.Net` is the official low-level .NET client of Elasticsearch.

I am using both of them.

arekpalinski replied:
Can you explain why we need / want to have both?

djordjedjukic replied:
Some things are easier to implement with the low-level client because we don't have strongly typed objects when indexing documents.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+        protected override int LoadInternal(IEnumerable<ElasticsearchIndexWithRecords> records, DocumentsOperationContext context, EtlStatsScope scope)
+        {
+            Uri[] nodes = Configuration.Connection.Nodes.Select(x => new Uri(x)).ToArray();
+            var pool = new StaticConnectionPool(nodes);
+            var settings = new ConnectionSettings(pool);
+            var client = new ElasticClient(settings);
```
Shouldn't this be global to the process, not created on each invocation of the `Load()`?

djordjedjukic replied:
I agree. I will change that.

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+            
+            foreach (var index in records)
+            {
+                StringBuilder deleteQuery = new StringBuilder();
```
Maybe reuse this to avoid allocating a lot of memory each time?

djordjedjukic replied:
I agree. I will change that.

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+                    .Query(q => q
+                        .Match(p => p
+                            .Field(index.IndexIdProperty)
+                            .Query($"{deleteQuery}"))
```
Is there a better way? I don't like building the query using string concat.
For that matter, you can just call `deleteQuery.ToString()` here.

Can we use an array rather than build a string?
What happens if the document id contains a ,` ?

djordjedjukic replied:
I will check this, and get back to you with the answer.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+                
+                if (deleteResponse.ServerError != null)
+                {
+                    throw new Exception($"ServerError: {deleteResponse.ServerError.Error}");
```
Don't throw `Exception`.
Provide more details here, such as the documents that we tried to delete, anything else that we can provide.

djordjedjukic replied:
We are deleting multiple documents in a single request. I will check does the response have those details.

When you say `don't throw an exception`, do you mean don't throw it at all, or just don't throw it like it is now? @ayende

ayende replied:
You are throwing a `Systme.Exception`, throw something more meaningful.
And if there are multiple documents, write all their ids.

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+                
+                if (deleteResponse.OriginalException != null)
+                {
+                    throw new Exception($"OriginalException: {deleteResponse.OriginalException.Message}");
```
Same as above

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+
+                statsCounter += (int)deleteResponse.Deleted;
+
+                foreach (ElasticsearchItem insert in index.Inserts)
```
Isn't there a way to do this that doesn't run one document at a time? That can be quite expensive.

djordjedjukic replied:
Yes, I didn't find any obvious way to send it in a single request for unstructured data (we are sending JSON).
I left this for the end, to try to optimize this part.

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,165 @@
+
+                    if (response.Success == false)
+                    {
+                        throw new Exception(response.OriginalException.Message);
```
Same - error handling ergonomics matter here.

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchIndexWriterSimulator.cs:
```diff
@@ -0,0 +1,69 @@
+                .Append("\"")
+                .Append(":")
+                .Append("\"")
+                .Append(deleteQuery)
```
Not handling escape properly

Can we generate that using proper JSON construction rather than by string concat?

arekpalinski replied:
Yes, it's gonna be changed to use DynamicJsonValue and generate JSON using it

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchIndexWriterSimulator.cs:
```diff
@@ -0,0 +1,69 @@
+            var result = new List<string>();
+
+            foreach (var item in elasticsearchItems)
+            {
+                var sb = new StringBuilder("POST ")
+                    .Append(indexName)
+                    .AppendLine("/_doc")
+                    .AppendLine("{")
+                    .AppendLine(item.Property.RawValue.ToString().Replace("{", "").Replace("}", ""))
+                    .AppendLine("}");
+
```
Why are you doing string replace here?

arekpalinski replied:
@djordjedjukic do we need it? why not just drop `.AppendLine("{")` and `.AppendLine("}");`  ?

djordjedjukic replied:
It can be done like that, I wanted to keep the string pretty formated. I guess that I should rewrite this with DynamicJsonValue  also?

djordjedjukic replied:
done

ayende commented on src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs:
```diff
@@ -1013,6 +1013,37 @@ public async ValueTask WriteDatabaseRecordAsync(DatabaseRecord databaseRecord, S
+                    progress.ElasticsearchEtlsUpdated = true;
+                }
+                
+                if (databaseRecord.ElasticsearchConnectionStrings.Count > 0 && databaseRecordItemType.HasFlag(DatabaseRecordItemType.ElasticsearchConnectionStrings))
```
Shouldn't you put the connection strings first, then create the task?

ayende commented on test/SlowTests/Server/Documents/ETL/ConnectionStringTests.cs:
```diff
@@ -39,6 +40,15 @@ public void CanAddAndRemoveConnectionStrings()
+                var elasticsearchConnectionString = new ElasticsearchConnectionString
+                {
+                    Name = "ElasticsearchConnectionString",
+                    Nodes = new[]{"http://127.0.0.1:8080" },
```
We usually use `8080` for RavenDB, which can be confusing.

djordjedjukic replied:
I will change that. The default for Elasticsearch is 9200

ayende commented on test/SlowTests/Server/Documents/ETL/Elasticsearch/ElasticsearchEtlTests.cs:
```diff
@@ -0,0 +1,871 @@
+        }
+
+        protected const string defaultScript = @"
+var orderData = {
```
Does this make sense in terms of Elastic usage? This is the script we use for SQL, no?

djordjedjukic replied:
It is. I reused it for testing `Parent-Child` relations.

ayende commented on test/SlowTests/Server/Documents/ETL/Elasticsearch/ElasticsearchEtlTests.cs:
```diff
@@ -0,0 +1,871 @@
+                    session.SaveChanges();
+                }
+
+                var node = new Uri("http://localhost:9200");
```
Let's move the connection to a constant or some known location, rather than in here

ayende commented on test/SlowTests/Server/Documents/ETL/Elasticsearch/ElasticsearchEtlTests.cs:
```diff
@@ -0,0 +1,871 @@
+            }
+        }
+
+        protected void SetupElasticEtl(DocumentStore store, string script, IEnumerable<string> collections = null, bool applyToAllDocuments = false)
```
How do we cleanup Elastic between calls to the tests?

djordjedjukic replied:
We can delete indexes after every test. Is that sound ok? @ayende

ayende replied:
Yes, that should do it.

ayende reviewed (changes requested):
Looks very nice, I had some comments

djordjedjukic committed:
rename from Elasticsearch => ElasticSearch

djordjedjukic committed:
export first connection strings the etls

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Rebase + Connection Strings View (initial work)

djordjedjukic committed:
authentication for elasticsearch

djordjedjukic committed:
remove unused code

djordjedjukic committed:
Merge remote-tracking branch 'djordjedjukic-ravendb/feature/13041' into feature/13041

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchHelper.cs:
```diff
@@ -0,0 +1,36 @@
+            }
+            else if (connectionString.Authentication.CertificateAuth != null)
+            {
+                // add certificates
```
TODO?

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Edit Task View + List View - Initial Work

djordjedjukic committed:
certificates auth

djordjedjukic committed:
certificates auth

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Add export & import + some fixes

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Fix Progress NRE

ml054 commented on src/Raven.Studio/typescript/models/database/settings/connectionStringElasticSearchEtlModel.ts:
```diff
@@ -0,0 +1,120 @@
+        };
+    }
+
+    // removeElasticUrl(url: discoveryUrl) {
```
?

ml054 commented on src/Raven.Studio/typescript/models/database/settings/connectionStringElasticSearchEtlModel.ts:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    addDiscoveryUrlWithBlink() {
+        if ( !_.find(this.nodesUrls(), x => x.discoveryUrlName() === this.inputUrl().discoveryUrlName())) {
```
remove space before `!`

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/settings/connectionStringElasticSearchEtlModel.ts:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    testConnection(urlToTest: discoveryUrl) : JQueryPromise<Raven.Server.Web.System.NodeConnectionTestResult> {
+        // return new testClusterNodeConnectionCommand(urlToTest.discoveryUrlName(), this.database(), false)
```
?

ml054 commented on src/Raven.Studio/typescript/models/database/settings/connectionStringElasticSearchEtlModel.ts:
```diff
@@ -0,0 +1,120 @@
+            });
+    }
+
+    saveConnectionString(db: database) : JQueryPromise<void> {
```
remove space before `:`

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/models/database/tasks/ongoingTaskElasticSearchEtlEditModel.ts:
```diff
@@ -0,0 +1,114 @@
+    protected initializeObservables() {
+        super.initializeObservables();
+        
+        // todo .. script dirty ?? check...
```
?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/settings/connectionStrings.ts:
```diff
@@ -83,6 +87,11 @@ class connectionStrings extends viewModelBase {
                 return olapEtl.dirtyFlag().isDirty();
             }
 
+            const ElasticEtl = this.editedElasticSearchEtlConnectionString();
```
lower case variable name

Danielle9897 replied:
done

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editElasticSearchEtlTask.ts:
```diff
@@ -0,0 +1,776 @@
+    }
+
+    runTest() {
+        // const testValid = viewHelpers.isValid(this.validationGroup, true);
+        // const parentValid = this.validateParent();
+        //
+        // if (testValid && parentValid) {
+        //     this.spinners.test(true);
+        //
+        //     const dto = {
+        //         DocumentId: this.documentId(),
+        //         IsDelete: this.testDelete(),
+        //         PerformRolledBackTransaction: this.performRolledBackTransaction(),
+        //         Configuration: this.configurationProvider(),
+        //         Connection: this.connectionProvider()
+        //     } as Raven.Server.Documents.ETL.Providers.SQL.RelationalWriters.TestSqlEtlScript;
+        //
+        //     eventsCollector.default.reportEvent("sql-etl", "test-replication");
+        //
+        //     new testSqlReplicationCommand(this.db(), dto)
+        //         .execute()
+        //         .done((testResult: Raven.Server.Documents.ETL.Providers.SQL.Test.SqlEtlTestScriptResult) => {
+        //             this.testResults(_.flatMap(testResult.Summary, x => x.Commands));
+        //             this.debugOutput(testResult.DebugOutput);
+        //             this.loadErrors(testResult.LoadErrors);
+        //             this.transformationErrors(testResult.TransformationErrors);
+        //
+        //             if (this.warningsCount()) {
+        //                 $('.test-container a[href="#warnings"]').tab('show');
+        //             } else {
+        //                 $('.test-container a[href="#testResults"]').tab('show');
+        //             }
+        //
+        //             this.testAlreadyExecuted(true);
+        //         })
+        //         .always(() => this.spinners.test(false));
+        // }
+    }
```
TODO

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editElasticSearchEtlTask.ts:
```diff
@@ -0,0 +1,776 @@
+    possibleMentors = ko.observableArray<string>([]);
+    elasticSearchEtlConnectionStringsNames = ko.observableArray<string>([]);
+
+    //connectionStringDefined: KnockoutComputed<boolean>;
```
?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editElasticSearchEtlTask.ts:
```diff
@@ -0,0 +1,776 @@
+        save: ko.observable<boolean>(false)
+    };
+    
+    // needed ??
```
?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editElasticSearchEtlTask.ts:
```diff
@@ -0,0 +1,776 @@
+    
+    collectionNames: KnockoutComputed<string[]>;
+
+    //showAdvancedOptions = ko.observable<boolean>(false);
```
?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/editElasticSearchEtlTask.ts:
```diff
@@ -0,0 +1,776 @@
+            this.editedElasticSearchEtl().connectionStringName(null);
+        }
+        
+        // this.connectionStringDefined = ko.pureComputed(() => {
+        //     const editedEtl = this.editedElasticSearchEtl();
+        //     if (this.createNewConnectionString()) {
+        //         return !!this.newConnectionString().connectionString(); // ???
+        //     } else {
+        //         return !!editedEtl.connectionStringName();
+        //     }
+        // });
```
?

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/tasks/ongoingTasks.ts:
```diff
@@ -520,7 +543,16 @@ class ongoingTasks extends viewModelBase {
 
     showItemPreview(item: ongoingTaskListModel, scriptName: string) {
-        const type: Raven.Client.Documents.Operations.ETL.EtlType = item.taskType() === "RavenEtl" ? "Raven" : item.taskType() === "SqlEtl" ? "Sql" : "Olap";
+        //const type: Raven.Client.Documents.Operations.ETL.EtlType = item.taskType() === "RavenEtl" ? "Raven" : item.taskType() === "SqlEtl" ? "Sql" : "Olap";
```
?

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Small fixes

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Stats View

djordjedjukic committed:
RavenDB-13041 connection string serialization

djordjedjukic committed:
RavenDB-13041 index name validation

djordjedjukic committed:
RavenDB-13041 missing code in smuggler

djordjedjukic committed:
RavenDB-13041 custom exception

djordjedjukic committed:
RavenDB-13041 use credentials auth in tests

djordjedjukic committed:
Merge remote-tracking branch 'djordjedjukic-ravendb/feature/13041' into feature/13041

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Add Authentication to Connection String

djordjedjukic committed:
RavenDB-13041 create client in LoadInteral metod

djordjedjukic committed:
RavenDB-13041 remove unused code

djordjedjukic committed:
RavenDB-13041 rewrite index name validator

djordjedjukic committed:
RavenDB-13041 serialize Auth to null if doesn't exist

djordjedjukic committed:
Merge remote-tracking branch 'djordjedjukic-ravendb/feature/13041' into feature/13041

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Test Script Area

djordjedjukic committed:
RavenDB-13041 test elasticsearch connection string

djordjedjukic committed:
RavenDB-13041 modify url for test conn string

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Test Elastic Node Connection

djordjedjukic committed:
RavenDB-13041 rework test connection string

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+    
+    public class Authentication
+    {
+        public ApiKeyAuth ApiKeyAuth { get; set; }
```
property should be `ApiKey` ?

djordjedjukic replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+    public class Authentication
+    {
+        public ApiKeyAuth ApiKeyAuth { get; set; }
+        public BasicAuth BasicAuth { get; set; }
```
property should be `Basic` ?

djordjedjukic replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+    {
+        public ApiKeyAuth ApiKeyAuth { get; set; }
+        public BasicAuth BasicAuth { get; set; }
+        public CertificateAuth CertificateAuth { get; set; }
```
property should be `Certificate` ?

djordjedjukic replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+        public CertificateAuth CertificateAuth { get; set; }
+    }
+
+    public class ApiKeyAuth
```
use full name - `ApiKeyAuthentication`

djordjedjukic replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+        public string ApiKey { get; set; }
+    }
+    
+    public class BasicAuth
```
use full name - `BasicAuthentication`

djordjedjukic replied:
done

ppekrol commented on src/Raven.Client/Documents/Operations/ETL/Elasticsearch/ElasticsearchConnectionString.cs:
```diff
@@ -0,0 +1,107 @@
+        public string Password { get; set; }
+    }
+    
+    public class CertificateAuth
```
use full name - CertificateAuthentication

djordjedjukic replied:
done

ppekrol commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchLoadException.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System;
+
+namespace Raven.Server.Documents.ETL.Providers.ElasticSearch
```
move it to Raven.Server.Exceptions.ETL.ElasitcSearch ?

djordjedjukic replied:
done

ppekrol commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -146,6 +146,13 @@
     <PackageReference Include="NuGet.Protocol" Version="5.11.0" />
     <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="3.21.3" />
+    <PackageReference Include="NEST" Version="7.13.0" />
+    <PackageReference Include="Npgsql" Version="5.0.5" />
```
merge issue, should be removed

djordjedjukic replied:
done

arekpalinski commented on src/Raven.Server/Documents/ETL/EtlLoader.cs:
```diff
@@ -229,6 +240,15 @@ private void HandleChangesSubscriptions()
                         break;
+                    case EtlType.ElasticSearch:
+                        elasticsearchConfig = config as ElasticSearchEtlConfiguration;
+                        if (_databaseRecord.ElasticSearchConnectionStrings.TryGetValue(config.ConnectionStringName, out var elasticsearchConnection))
```
Minor: elasticsearchConnection ->elasticSearchConnection. In general for consistency let's do the case sensitive search for "elasticsearch" / "Elasticsearch" and rename to "elasticSearch" / "ElasticSearch"

djordjedjukic replied:
done

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchLoadException.cs:
```diff
@@ -0,0 +1,21 @@
+﻿using System;
```
I have just noticed that the directory name is still "Elasticsearch". See  `src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticSearchLoadException.cs`

djordjedjukic replied:
done

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchDocumentTransformer.cs:
```diff
@@ -0,0 +1,120 @@
+                    ThrowIndexNotDefinedInConfig(indexName);
+
+                _indexes[indexName] =
+                    index = new ElasticSearchIndexWithRecords(elasticIndex);
```
Strange formatting

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/ElasticsearchEtl.cs:
```diff
@@ -0,0 +1,182 @@
+                
+                if (deleteResponse.OriginalException != null)
+                {
+                    throw new ElasticSearchLoadException($"Index {index}; Documents IDs: {deleteQuery}; error: {deleteResponse.OriginalException}");
```
Maybe move the exception throwing to a dedicated method and just pass `index` and `DebugInformation / ServerError.Error / OriginalException`

arekpalinski commented on src/Raven.Server/Documents/ETL/Providers/Elasticsearch/Handlers/ElasticSearchEtlConnectionHandler.cs:
```diff
@@ -0,0 +1,59 @@
+
+            ElasticClient client = ElasticSearchHelper.CreateClient(new ElasticSearchConnectionString { Nodes = new[] { url }, Authentication = authentication });
+
+            PingResponse pingResult = await client.PingAsync();
```
I think it should be wrapped by `try / catch` same like in SQL ETL:

https://github.com/ravendb/ravendb/blob/v5.2/src/Raven.Server/Documents/ETL/Providers/SQL/Handlers/SqlEtlHandler.cs#L44-L59

djordjedjukic replied:
done

arekpalinski commented on src/Raven.Server/Raven.Server.csproj:
```diff
@@ -146,6 +146,13 @@
+    <PackageReference Include="NuGet.Resolver" Version="5.8.1" />
+    <PackageReference Include="NuGet.PackageManagement" Version="5.8.1" />
+    <PackageReference Include="NuGet.Protocol" Version="5.8.1" />
+    <PackageReference Include="Oracle.ManagedDataAccess.Core" Version="3.21.1" />
```
I think we only need one like here:

`<PackageReference Include="NEST" Version="7.13.0" />`

arekpalinski replied:
Note that the rest is already specified above

djordjedjukic replied:
done

djordjedjukic committed:
RavenDB-13041 elasticsearch => elasticSearch

djordjedjukic committed:
RavenDB-13041 rename authentication properties

djordjedjukic committed:
RavenDB-13041 cover test connection with try/catch

djordjedjukic committed:
RavenDB-13041 fix merge issues

djordjedjukic committed:
RavenDB-13041 rise warning when auth is missing

djordjedjukic committed:
RavenDB-13041 elastic etl exception moved

djordjedjukic committed:
RavenDB-13041 ignore not existing index error on delete

djordjedjukic committed:
RavenDB-13041 use DJV instead of string

Danielle9897 committed:
RavenDB-13041 Fix Studio Compilation

djordjedjukic committed:
RavenDB-13041 tests refactoring

djordjedjukic committed:
Merge remote-tracking branch 'djordjedjukic-ravendb/feature/13041' into feature/13041

djordjedjukic committed:
RavenDB-13041 code cleanup

ppekrol committed:
Merge branch 'v5.2' of https://github.com/ravendb/ravendb into feature/13041

arekpalinski committed:
RavenDB-13041 Added missing code to handle ETL status update for ElasticSearch. Refactoring and fixing tests

arekpalinski committed:
RavenDB-13041 Added RequiresElasticSearchFactAttribute. Temporarily skip license verification agains ES ETL feature

arekpalinski committed:
RavenDB-13041
- introducing InsertOnlyMode
- using terms query instead of match
- forcing index refresh before doing DeleteByQuery instead of using Refresh.WaitFor option when inserting data
- creating indexes explicitly if then don't exists - IMPORTANT: the IndexIdProperty must not be analyzed - we need to search by exact match when doing deletions (known issue: we need to lowercase when doing inserts)
- added new test which revealed the issue with deletions - it has more than single document to process

Danielle9897 committed:
RavenDB-17147 Elastic Search ETL - Studio: Add Icon + Insert Only Mode

arekpalinski committed:
RavenDB-13041 Rename ElasticsearchEtl to ElasticSearchEtl

arekpalinski committed:
RavenDB-13041 Rename ElasticsearchEtlTests to ElasticSearchEtlTests

arekpalinski committed:
RavenDB-13041 Making sure that we're sending lower cased document ids so the we're able to delete items when sending DeleteByQuery on a field which is not analyzed (in default index definition we mark the field as Keyword so they are serchable only by exact match)

arekpalinski committed:
RavenDB-13041 Updating the ES simulator so the test scripts will reflect the current implementation of Load method

arekpalinski committed:
RavenDB-13041 No need to check if DebugInfo is null or empty

arekpalinski committed:
RavenDB-13041 Fixing the access to test endpoints for db admin. Fixing failing test related to adding new fields to db record.

arekpalinski committed:
RavenDB-13041 Renamed ElasticsearchItem.cs to ElasticSearchItem.cs

Danielle9897 committed:
RavenDB-17147 Fix studio compilation + verify DatabaseAdmin Clearance

arekpalinski committed:
RavenDB-13041 Fixing and issue with putting empty commands when using test functionality

arekpalinski committed:
RavenDB-13041 Fixing import of Elastic Search task on a database import

Danielle9897 committed:
RavenDB-17147 Fix Sample Script

arekpalinski committed:
RavenDB-13041 Inserting all items for a given index in a single bulk request instead of doing a request per a document

arekpalinski committed:
RavenDB-13041 Making sure that insert only mode is respected in test script feature

arekpalinski committed:
RavenDB-13041 Renamed ElasticsearchProperty to ElasticSearchProperty.cs

arekpalinski committed:
RavenDB-13041 Renamed ElasticsearchDocumentTransformer.cs to ElasticSearchDocumentTransformer.cs

arekpalinski committed:
RavenDB-13041Renamed ElasticsearchEtlConfiguration.cs to ElasticsearchEtlConfiguration.cs

Danielle9897 committed:
RavenDB-17147 Fix \n in tests results for FireFox

Danielle9897 committed:
RavenDB-17147 Fix \n in tests results for FireFox - Fix Review Comment

ppekrol committed:
Merge branch 'v5.3' of https://github.com/ravendb/ravendb into feature/13041
